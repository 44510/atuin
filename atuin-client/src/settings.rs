use std::fs::{create_dir_all, File};
use std::io::prelude::*;
use std::path::{Path, PathBuf};

use chrono::prelude::*;
use chrono::Utc;
use config::{Config, Environment, File as ConfigFile};
use eyre::{eyre, Context, Result};
use parse_duration::parse;

pub const HISTORY_PAGE_SIZE: i64 = 100;

#[derive(Clone, Debug, Deserialize, Copy)]
pub enum SearchMode {
    #[serde(rename = "prefix")]
    Prefix,

    #[serde(rename = "fulltext")]
    FullText,

    #[serde(rename = "fuzzy")]
    Fuzzy,
}

// FIXME: Can use upstream Dialect enum if https://github.com/stevedonovan/chrono-english/pull/16 is merged
#[derive(Clone, Debug, Deserialize, Copy)]
pub enum Dialect {
    #[serde(rename = "us")]
    Us,

    #[serde(rename = "uk")]
    Uk,
}

impl From<Dialect> for chrono_english::Dialect {
    fn from(d: Dialect) -> chrono_english::Dialect {
        match d {
            Dialect::Uk => chrono_english::Dialect::Uk,
            Dialect::Us => chrono_english::Dialect::Us,
        }
    }
}

#[derive(Clone, Debug, Deserialize)]
pub struct Settings {
    pub dialect: Dialect,
    pub auto_sync: bool,
    pub sync_address: String,
    pub sync_frequency: String,
    pub db_path: String,
    pub key_path: String,
    pub session_path: String,
    pub search_mode: SearchMode,

    // This is automatically loaded when settings is created. Do not set in
    // config! Keep secrets and settings apart.
    #[serde(skip)]
    pub session_token: Option<String>,
}

impl Settings {
    pub fn save_sync_time() -> Result<()> {
        let data_dir = atuin_common::utils::data_dir();
        let data_dir = data_dir.as_path();

        let sync_time_path = data_dir.join("last_sync_time");

        std::fs::write(sync_time_path, Utc::now().to_rfc3339())?;

        Ok(())
    }

    pub fn last_sync() -> Result<chrono::DateTime<Utc>> {
        let data_dir = atuin_common::utils::data_dir();
        let data_dir = data_dir.as_path();

        let sync_time_path = data_dir.join("last_sync_time");

        if !sync_time_path.exists() {
            return Ok(Utc.ymd(1970, 1, 1).and_hms(0, 0, 0));
        }

        let time = std::fs::read_to_string(sync_time_path)?;
        let time = chrono::DateTime::parse_from_rfc3339(time.as_str())?;

        Ok(time.with_timezone(&Utc))
    }

    pub fn should_sync(&self) -> Result<bool> {
        let session_path = atuin_common::utils::data_dir().join("session");

        if !self.auto_sync || !session_path.exists() {
            return Ok(false);
        }

        match parse(self.sync_frequency.as_str()) {
            Ok(d) => {
                let d = chrono::Duration::from_std(d).unwrap();
                Ok(Utc::now() - Settings::last_sync()? >= d)
            }
            Err(e) => Err(eyre!("failed to check sync: {}", e)),
        }
    }

    pub fn new() -> Result<Self> {
        let config_dir = atuin_common::utils::config_dir();

        let data_dir = atuin_common::utils::data_dir();

        create_dir_all(&config_dir)
            .wrap_err_with(|| format!("could not create dir {:?}", config_dir))?;
        create_dir_all(&data_dir)
            .wrap_err_with(|| format!("could not create dir {:?}", data_dir))?;

        let mut config_file = if let Ok(p) = std::env::var("ATUIN_CONFIG_DIR") {
            PathBuf::from(p)
        } else {
            let mut config_file = PathBuf::new();
            config_file.push(config_dir);
            config_file
        };

        config_file.push("config.toml");

        let db_path = data_dir.join("history.db");
        let key_path = data_dir.join("key");
        let session_path = data_dir.join("session");

        let mut s = Config::builder()
            .set_default("db_path", db_path.to_str())?
            .set_default("key_path", key_path.to_str())?
            .set_default("session_path", session_path.to_str())?
            .set_default("dialect", "us")?
            .set_default("auto_sync", true)?
            .set_default("sync_frequency", "1h")?
            .set_default("sync_address", "https://api.atuin.sh")?
            .set_default("search_mode", "prefix")?;

        if config_file.exists() {
            s = s.add_source(ConfigFile::with_name(config_file.to_str().unwrap()));
        } else {
            let example_config = include_bytes!("../config.toml");
            let mut file = File::create(config_file).wrap_err("could not create config file")?;
            file.write_all(example_config)
                .wrap_err("could not write default config file")?;
        }

        s = s.add_source(Environment::with_prefix("atuin").separator("_"));

        let mut settings: Self = s
            .build()
            .wrap_err("could not process config builder")?
            .try_into()
            .map_err(|e| eyre!("failed to deserialize: {}", e))?;

        // all paths should be expanded
        settings.db_path = shellexpand::full(&settings.db_path)?.into_owned();
        settings.key_path = shellexpand::full(&settings.key_path)?.into_owned();
        settings.session_path = shellexpand::full(&settings.session_path)?.into_owned();

        // Finally, set the auth token
        if Path::new(&settings.session_path).exists() {
            let token = std::fs::read_to_string(&settings.session_path)?;
            settings.session_token = Some(token.trim().to_owned());
        } else {
            settings.session_token = None;
        }

        Ok(settings)
    }
}
