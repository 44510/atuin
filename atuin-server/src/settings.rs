use std::fs::{create_dir_all, File};
use std::io::prelude::*;
use std::path::PathBuf;

use config::{Config, Environment, File as ConfigFile};
use eyre::{Context, Result};

pub const HISTORY_PAGE_SIZE: i64 = 100;

#[derive(Clone, Debug, Deserialize)]
pub struct Settings {
    pub host: String,
    pub port: u16,
    pub db_uri: String,
    pub open_registration: bool,
}

impl Settings {
    pub fn new() -> Result<Self> {
        let config_dir = atuin_common::utils::config_dir();
        let config_dir = config_dir.as_path();

        create_dir_all(config_dir)?;

        let mut config_file = if let Ok(p) = std::env::var("ATUIN_CONFIG_DIR") {
            PathBuf::from(p)
        } else {
            let mut config_file = PathBuf::new();
            config_file.push(config_dir);
            config_file
        };

        config_file.push("server.toml");

        // create the config file if it does not exist

        let mut s = Config::builder();

        if config_file.exists() {
            s = s.add_source(ConfigFile::with_name(config_file.to_str().unwrap()));
        } else {
            let example_config = include_bytes!("../server.toml");
            let mut file = File::create(config_file)?;
            file.write_all(example_config)?;
        }

        let s = s
            .set_default("host", "127.0.0.1")?
            .set_default("port", 8888)?
            .set_default("open_registration", false)?
            .set_default("db_uri", "default_uri")?
            .add_source(Environment::with_prefix("atuin").separator("_"));

        s.build()
            .wrap_err("failed to create config parser")?
            .try_into()
            .wrap_err("failed to deserialize")
    }
}
