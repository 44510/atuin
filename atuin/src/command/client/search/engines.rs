use async_trait::async_trait;
use atuin_client::{
    database::{Context, Database},
    history::History,
    settings::{FilterMode, SearchMode},
};
use chrono::{DateTime, Utc};
use eyre::Result;

use super::cursor::Cursor;

pub mod db;
pub mod skim;

#[cfg(test)]
pub mod test;

pub fn engine(search_mode: SearchMode) -> Box<dyn SearchEngine> {
    match search_mode {
        SearchMode::Skim => Box::new(skim::Search::new()) as Box<_>,
        mode => Box::new(db::Search(mode)) as Box<_>,
    }
}

pub struct SearchState {
    pub input: Cursor,
    pub filter_mode: FilterMode,
    pub context: Context,
}

#[async_trait]
pub trait SearchEngine: Send + Sync + 'static {
    async fn full_query_since(
        &mut self,
        state: &SearchState,
        db: &mut dyn Database,
        now: DateTime<Utc>,
    ) -> Result<Vec<History>>;

    async fn full_query(
        &mut self,
        state: &SearchState,
        db: &mut dyn Database,
    ) -> Result<Vec<History>> {
        self.full_query_since(state, db, Utc::now()).await
    }

    async fn query(&mut self, state: &SearchState, db: &mut dyn Database) -> Result<Vec<History>> {
        if state.input.as_str().is_empty() {
            Ok(db
                .list(state.filter_mode, &state.context, Some(200), true)
                .await?
                .into_iter()
                .collect::<Vec<_>>())
        } else {
            self.full_query(state, db).await
        }
    }
}
