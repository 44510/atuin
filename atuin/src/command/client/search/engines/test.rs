use atuin_client::{database::{Database, Context}, history::History, settings::FilterMode};
use chrono::{Utc, TimeZone};

use super::{SearchEngine, SearchState};

async fn entries(since: chrono::DateTime<chrono::Utc>) -> atuin_client::database::Sqlite {
    use atuin_client::database::Sqlite;
    use chrono::Duration;

    let mut db = Sqlite::new("sqlite::memory:").await.unwrap();

    let mut history: [History; 2] = [
        History::import()
            .timestamp(since - Duration::days(2))
            .command("docker run -e POSTGRES_USER=atuin -e POSTGRES_PASSWORD=pass -e POSTGRES_DB=atuin -p 5432:5432 -d --rm postgres:14-alpine")
            .session("1")
            .cwd("/Users/conrad/code/atuin")
            .hostname("host1:conrad")
            .build()
            .into(),
        History::import()
            .timestamp(since - Duration::days(2))
            .command("cargo sqlx prepare --database-url postgresql://user:pass@localhost/store")
            .session("1")
            .cwd("/Users/conrad/code/atuin")
            .hostname("host1:conrad")
            .build()
            .into(),
    ];

    for (i, h) in history.iter_mut().enumerate() {
        h.id = i.to_string();
    }

    db.save_bulk(&history).await.unwrap();

    db
}

pub async fn docker_postgres(mut search: impl SearchEngine) -> Vec<History> {
    let now = Utc.ymd(2023, 6, 18).and_hms(11, 14, 16);
    let mut db = entries(now).await;

    let state = SearchState {
        input: "docker postgres".to_owned().into(),
        filter_mode: FilterMode::Global,
        context: Context {
            session: "1".into(),
            cwd: "/Users/conrad/code/atuin".into(),
            hostname: "host1:conrad".into(),
            host_id: String::new(),
        },
    };

    search.full_query(&state, &mut db).await.unwrap()
}

pub async fn postgres(mut search: impl SearchEngine) -> Vec<History> {
    let now = Utc.ymd(2023, 6, 18).and_hms(11, 14, 16);
    let mut db = entries(now).await;

    let state = SearchState {
        input: "postgres".to_owned().into(),
        filter_mode: FilterMode::Global,
        context: Context {
            session: "1".into(),
            cwd: "/Users/conrad/code/atuin".into(),
            hostname: "host1:conrad".into(),
            host_id: String::new(),
        },
    };

    search.full_query(&state, &mut db).await.unwrap()
}
