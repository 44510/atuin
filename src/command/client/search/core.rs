use std::{
    ops::{ControlFlow, Deref},
    sync::Arc,
};

use eyre::Result;
use semver::Version;

use atuin_client::{
    database::Context,
    database::{current_context, Database},
    history::History,
    settings::{ExitMode, FilterMode, SearchMode, Settings},
};
use skim::SkimItem;

use super::{cursor::Cursor, history_list::ListState};

pub struct State<DB: Database> {
    pub db: DB,
    pub results_state: ListState,
    pub history: Vec<Arc<HistoryWrapper>>,
    pub history_count: i64,
    pub settings: Settings,
    pub update_needed: Option<Version>,

    pub search: SearchState,

    // only allocated if using skim
    pub all_history: Vec<Arc<HistoryWrapper>>,
}

pub struct SearchState {
    pub input: Cursor,
    pub filter_mode: FilterMode,
    pub search_mode: SearchMode,
    /// Store if the user has _just_ changed the search mode.
    /// If so, we change the UI to show the search mode instead
    /// of the filter mode until user starts typing again.
    switched_search_mode: bool,
    pub context: Context,
}

pub struct HistoryWrapper {
    history: History,
    pub count: i32,
}
impl Deref for HistoryWrapper {
    type Target = History;

    fn deref(&self) -> &Self::Target {
        &self.history
    }
}
impl SkimItem for HistoryWrapper {
    fn text(&self) -> std::borrow::Cow<str> {
        std::borrow::Cow::Borrowed(self.history.command.as_str())
    }
}

pub struct BatchGuard<DB: Database> {
    initial_input: String,
    initial_filter_mode: FilterMode,
    inner: State<DB>,
}

#[derive(Clone)]
pub enum Line {
    Up,
    Down,
}
#[derive(Clone)]
pub enum For {
    Page,
    SingleLine,
}
#[derive(Clone)]
pub enum Towards {
    Left,
    Right,
}
#[derive(Clone)]
pub enum To {
    Word,
    Char,
    Edge,
}

#[derive(Clone)]
pub enum Event {
    Input(char),
    InputStr(String),
    Selection(Line, For),
    Cursor(Towards, To),
    Delete(Towards, To),
    Clear,
    Exit,
    UpdateNeeded(Version),
    Cancel,
    SelectN(u32),
    CycleFilterMode,
    CycleSearchMode,
}

impl<DB: Database> State<DB> {
    /// Create a new core state
    pub async fn new(query: &[String], settings: Settings, db: DB) -> Result<Self> {
        let mut input = Cursor::from(query.join(" "));
        // Put the cursor at the end of the query by default
        input.end();

        let mut core = Self {
            history_count: db.history_count().await?,
            results_state: ListState::default(),
            update_needed: None,
            history: Vec::new(),
            db,
            all_history: Vec::new(),
            search: SearchState {
                input,
                context: current_context(),
                filter_mode: if settings.shell_up_key_binding {
                    settings
                        .filter_mode_shell_up_key_binding
                        .unwrap_or(settings.filter_mode)
                } else {
                    settings.filter_mode
                },
                search_mode: settings.search_mode,
                switched_search_mode: false,
            },
            settings,
        };
        core.refresh_query().await?;
        Ok(core)
    }

    async fn refresh_query(&mut self) -> Result<()> {
        let i = self.search.input.as_str();
        let results = if i.is_empty() {
            self.db
                .list(
                    self.search.filter_mode,
                    &self.search.context,
                    Some(200),
                    true,
                )
                .await?
                .into_iter()
                .map(|history| HistoryWrapper { history, count: 1 })
                .map(Arc::new)
                .collect::<Vec<_>>()
        } else if self.search.search_mode == SearchMode::Skim {
            if self.all_history.is_empty() {
                self.all_history = self
                    .db
                    .all_with_count()
                    .await
                    .unwrap()
                    .into_iter()
                    .map(|(history, count)| HistoryWrapper { history, count })
                    .map(Arc::new)
                    .collect::<Vec<_>>();
            }

            super::skim_impl::fuzzy_search(&self.search, &self.all_history).await
        } else {
            self.db
                .search(
                    self.search.search_mode,
                    self.search.filter_mode,
                    &self.search.context,
                    i,
                    Some(200),
                    None,
                    None,
                )
                .await?
                .into_iter()
                .map(|history| HistoryWrapper { history, count: 1 })
                .map(Arc::new)
                .collect::<Vec<_>>()
        };

        self.results_state.select(0);
        self.history = results;
        Ok(())
    }

    fn handle(mut self, event: Event) -> ControlFlow<String, Self> {
        // reset the state, will be set to true later if user really did change it
        self.search.switched_search_mode = false;
        let len = self.history.len();
        match event {
            // moving the selection up and down
            Event::Selection(Line::Up, For::SingleLine) => {
                let i = self.results_state.selected() + 1;
                self.results_state.select(i.min(len - 1));
            }
            Event::Selection(Line::Down, For::SingleLine) => {
                let Some(i) = self.results_state.selected().checked_sub(1) else {
                    return ControlFlow::Break(String::new())
                };
                self.results_state.select(i);
            }
            Event::Selection(Line::Down, For::Page) => {
                let scroll_len =
                    self.results_state.max_entries() - self.settings.scroll_context_lines;
                let i = self.results_state.selected().saturating_sub(scroll_len);
                self.results_state.select(i);
            }
            Event::Selection(Line::Up, For::Page) => {
                let scroll_len =
                    self.results_state.max_entries() - self.settings.scroll_context_lines;
                let i = self.results_state.selected() + scroll_len;
                self.results_state.select(i.min(len - 1));
            }

            // moving the search cursor left and right
            Event::Cursor(Towards::Left, To::Char) => {
                self.search.input.left();
            }
            Event::Cursor(Towards::Right, To::Char) => self.search.input.right(),
            Event::Cursor(Towards::Left, To::Word) => self
                .search
                .input
                .prev_word(&self.settings.word_chars, self.settings.word_jump_mode),
            Event::Cursor(Towards::Right, To::Word) => self
                .search
                .input
                .next_word(&self.settings.word_chars, self.settings.word_jump_mode),
            Event::Cursor(Towards::Left, To::Edge) => self.search.input.start(),
            Event::Cursor(Towards::Right, To::Edge) => self.search.input.end(),

            // modifying the search
            Event::Input(c) => self.search.input.insert(c),
            Event::InputStr(s) => s.chars().for_each(|c| self.search.input.insert(c)),
            Event::Delete(Towards::Left, To::Word) => self
                .search
                .input
                .remove_prev_word(&self.settings.word_chars, self.settings.word_jump_mode),
            Event::Delete(Towards::Left, To::Char) => self.search.input.back(),
            Event::Delete(Towards::Left, To::Edge) => self.search.input.clear_from_start(),
            Event::Delete(Towards::Right, To::Word) => self
                .search
                .input
                .remove_next_word(&self.settings.word_chars, self.settings.word_jump_mode),
            Event::Delete(Towards::Right, To::Char) => self.search.input.remove(),
            Event::Delete(Towards::Right, To::Edge) => self.search.input.clear_to_end(),
            Event::Clear => self.search.input.clear(),

            // exiting
            Event::Cancel => return ControlFlow::Break(String::new()),
            Event::Exit => {
                return ControlFlow::Break(match self.settings.exit_mode {
                    ExitMode::ReturnOriginal => String::new(),
                    ExitMode::ReturnQuery => self.search.input.into_inner(),
                })
            }
            Event::SelectN(n) => {
                let i = self.results_state.selected().saturating_add(n as usize);
                return ControlFlow::Break(if i < self.history.len() {
                    self.search.input.into_inner()
                } else {
                    self.history.swap_remove(i).command.clone()
                });
            }

            // misc
            Event::UpdateNeeded(version) => self.update_needed = Some(version),
            Event::CycleSearchMode => {
                self.search.switched_search_mode = true;
                self.search.search_mode = self.search.search_mode.next(&self.settings);
            }
            Event::CycleFilterMode => {
                pub static FILTER_MODES: [FilterMode; 4] = [
                    FilterMode::Global,
                    FilterMode::Host,
                    FilterMode::Session,
                    FilterMode::Directory,
                ];
                let i = self.search.filter_mode as usize;
                let i = (i + 1) % FILTER_MODES.len();
                self.search.filter_mode = FILTER_MODES[i];
            }
        }
        ControlFlow::Continue(self)
    }

    /// Start a batch process of events
    pub fn start_batch(self) -> BatchGuard<DB> {
        BatchGuard {
            initial_input: self.search.input.as_str().to_owned(),
            initial_filter_mode: self.search.filter_mode,
            inner: self,
        }
    }

    /// Get the view of the state for rendering
    pub fn view(&mut self) -> View<'_> {
        View {
            history_count: self.history_count,
            search: &self.search,
            results_state: &mut self.results_state,
            update_needed: self.update_needed.as_ref(),
            history: &self.history,
        }
    }
}

impl<DB: Database> BatchGuard<DB> {
    /// Handle an event in the batch
    pub fn handle(mut self, event: Event) -> ControlFlow<String, Self> {
        match self.inner.handle(event) {
            ControlFlow::Continue(inner) => self.inner = inner,
            ControlFlow::Break(result) => return ControlFlow::Break(result),
        }
        ControlFlow::Continue(self)
    }

    /// Finish processing a batch of events, maybe refreshing the DB in the process
    pub async fn finish(self) -> Result<(State<DB>, bool)> {
        let Self {
            initial_input,
            initial_filter_mode,
            mut inner,
        } = self;
        let should_update = initial_input != inner.search.input.as_str()
            || initial_filter_mode != inner.search.filter_mode;
        if should_update {
            inner.refresh_query().await?;
        }

        Ok((inner, should_update))
    }
}

/// A view of the state for rendering
pub struct View<'a> {
    pub history_count: i64,
    pub results_state: &'a mut ListState,
    pub update_needed: Option<&'a Version>,
    pub history: &'a [Arc<HistoryWrapper>],

    pub search: &'a SearchState,
}
