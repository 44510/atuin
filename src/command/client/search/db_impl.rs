use std::sync::Arc;

use async_trait::async_trait;
use atuin_client::{database::Database, settings::SearchMode};
use eyre::Result;

use super::interactive::{HistoryWrapper, SearchEngine, SearchState};

pub struct Search(pub SearchMode);

#[async_trait]
impl SearchEngine for Search {
    async fn query(
        &mut self,
        state: &SearchState,
        db: &mut dyn Database,
    ) -> Result<Vec<Arc<HistoryWrapper>>> {
        Ok(db
            .search(
                self.0,
                state.filter_mode,
                &state.context,
                state.input.as_str(),
                Some(200),
                None,
                None,
            )
            .await?
            .into_iter()
            .map(|history| HistoryWrapper { history, count: 1 })
            .map(Arc::new)
            .collect::<Vec<_>>())
    }
}
