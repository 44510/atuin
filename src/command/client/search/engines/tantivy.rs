use std::sync::Arc;

use async_trait::async_trait;
use atuin_client::{
    database::Database,
    history::History,
    settings::FilterMode,
    tantivy::{index, schema, HistorySchema},
};
use chrono::{TimeZone, Utc};
use eyre::Result;
use tantivy::{
    collector::TopDocs,
    query::{BooleanQuery, ConstScoreQuery, FuzzyTermQuery, Occur, Query, TermQuery},
    schema::Value,
    Searcher, Term,
};

use super::{HistoryWrapper, SearchEngine, SearchState};

pub struct Search {
    schema: HistorySchema,
    searcher: Searcher,
}

impl Search {
    pub fn new() -> Result<Self> {
        let (hs, schema) = schema();
        let index = index(schema)?;

        let reader = index.reader()?;
        let searcher = reader.searcher();

        Ok(Self {
            schema: hs,
            searcher,
        })
    }
}

#[async_trait]
impl SearchEngine for Search {
    async fn full_query(
        &mut self,
        state: &SearchState,
        _: &mut dyn Database,
    ) -> Result<Vec<Arc<HistoryWrapper>>> {
        let mut queries = Vec::<(Occur, Box<dyn Query>)>::new();
        for term in state.input.as_str().split_whitespace() {
            let command =
                FuzzyTermQuery::new(Term::from_field_text(self.schema.command, term), 2, true);
            queries.push((Occur::Should, Box::new(command)));
        }

        match state.filter_mode {
            FilterMode::Global => {}
            FilterMode::Directory => {
                let cwd = TermQuery::new(
                    Term::from_field_text(self.schema.cwd, &state.context.cwd),
                    tantivy::schema::IndexRecordOption::Basic,
                );
                queries.push((
                    Occur::Must,
                    Box::new(ConstScoreQuery::new(Box::new(cwd), 1.0)),
                ));
            }
            FilterMode::Host => {
                let host = TermQuery::new(
                    Term::from_field_text(self.schema.hostname, &state.context.hostname),
                    tantivy::schema::IndexRecordOption::Basic,
                );
                queries.push((
                    Occur::Must,
                    Box::new(ConstScoreQuery::new(Box::new(host), 1.0)),
                ));
            }
            FilterMode::Session => {
                let session = TermQuery::new(
                    Term::from_field_text(self.schema.session, &state.context.session),
                    tantivy::schema::IndexRecordOption::Basic,
                );
                queries.push((
                    Occur::Must,
                    Box::new(ConstScoreQuery::new(Box::new(session), 1.0)),
                ));
            }
        }

        let query = BooleanQuery::new(queries);

        let top_docs = self.searcher.search(&query, &TopDocs::with_limit(200))?;

        let mut output = Vec::with_capacity(top_docs.len());

        for (_score, doc_address) in top_docs {
            let retrieved_doc = self.searcher.doc(doc_address)?;
            output.push(Arc::new(HistoryWrapper {
                history: History {
                    id: String::new(),
                    command: retrieved_doc
                        .get_all(self.schema.command)
                        .next()
                        .and_then(Value::as_text)
                        .unwrap()
                        .to_string(),
                    cwd: String::new(),
                    session: String::new(),
                    hostname: String::new(),
                    timestamp: retrieved_doc
                        .get_all(self.schema.timestamp)
                        .next()
                        .and_then(Value::as_date)
                        .and_then(|d| Utc.timestamp_millis_opt(d.into_timestamp_millis()).single())
                        .unwrap(),
                    duration: retrieved_doc
                        .get_all(self.schema.duration)
                        .next()
                        .and_then(Value::as_i64)
                        .unwrap(),
                    exit: retrieved_doc
                        .get_all(self.schema.exit)
                        .next()
                        .and_then(Value::as_i64)
                        .unwrap(),
                    deleted_at: None,
                },
                count: 1,
            }));
        }

        Ok(output)
    }
}
