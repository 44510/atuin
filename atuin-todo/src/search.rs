use atuin_common::record::Record;
use eyre::Result;
use tantivy::{
    directory::MmapDirectory,
    doc,
    schema::{Field, Schema, STORED, STRING, TEXT},
    DateTime, Index, IndexWriter, Term,
};

use crate::{record::TodoRecord, TodoId};

pub fn schema() -> (TodoSchema, Schema) {
    let mut schema_builder = Schema::builder();

    (
        TodoSchema {
            id: schema_builder.add_text_field("id", STRING | STORED),
            text: schema_builder.add_text_field("text", TEXT),
            timestamp: schema_builder.add_date_field("timestamp", STORED),
            tag: schema_builder.add_text_field("tag", TEXT),
        },
        schema_builder.build(),
    )
}

pub struct TodoSchema {
    pub id: Field,
    pub text: Field,
    pub timestamp: Field,
    pub tag: Field,
}

pub fn index(schema: Schema) -> Result<Index> {
    let data_dir = atuin_common::utils::data_dir().join("todo");
    let tantivy_dir = data_dir.join("tantivy");

    fs_err::create_dir_all(&tantivy_dir)?;
    let dir = MmapDirectory::open(tantivy_dir)?;

    Ok(Index::open_or_create(dir, schema)?)
}

pub fn write_record(
    writer: &mut IndexWriter,
    schema: &TodoSchema,
    record: &Record<TodoRecord>,
) -> Result<()> {
    let timestamp = DateTime::from_timestamp_nanos(record.timestamp as i64);
    let mut doc = doc!(
        schema.id => TodoId::from_uuid(record.id.0).to_string(),
        schema.text => record.data.text.clone(),
        schema.timestamp => timestamp,
    );
    for tag in record.data.tags.clone() {
        doc.add_field_value(schema.tag, tag)
    }
    writer.add_document(doc)?;

    if !record.data.updates.uuid().is_nil() {
        writer.delete_term(Term::from_field_text(
            schema.id,
            &record.data.updates.to_string(),
        ));
    }

    Ok(())
}
