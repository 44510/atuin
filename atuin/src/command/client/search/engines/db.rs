use async_trait::async_trait;
use atuin_client::{
    database::Database, database::OptFilters, history::History, settings::SearchMode,
};
use chrono::{DateTime, Utc};
use eyre::Result;

use super::{SearchEngine, SearchState};

pub struct Search(pub SearchMode);

#[async_trait]
impl SearchEngine for Search {
    async fn full_query_since(
        &mut self,
        state: &SearchState,
        db: &mut dyn Database,
        _now: DateTime<Utc>,
    ) -> Result<Vec<History>> {
        Ok(db
            .search(
                self.0,
                state.filter_mode,
                &state.context,
                state.input.as_str(),
                OptFilters {
                    limit: Some(200),
                    ..Default::default()
                },
            )
            .await?
            .into_iter()
            .collect::<Vec<_>>())
    }
}

#[cfg(test)]
mod tests {
    mod fuzzy {
        use atuin_client::settings::SearchMode;
        use insta::assert_debug_snapshot;

        use crate::command::client::search::engines::{self, db::Search};

        #[tokio::test]
        async fn docker_postgres() {
            assert_debug_snapshot!(engines::test::docker_postgres(Search(SearchMode::Fuzzy)).await);
        }

        #[tokio::test]
        async fn postgres() {
            assert_debug_snapshot!(engines::test::postgres(Search(SearchMode::Fuzzy)).await);
        }
    }

    mod full_text {
        use atuin_client::settings::SearchMode;
        use insta::assert_debug_snapshot;

        use crate::command::client::search::engines::{self, db::Search};

        #[tokio::test]
        async fn docker_postgres() {
            assert_debug_snapshot!(engines::test::docker_postgres(Search(SearchMode::FullText)).await);
        }

        #[tokio::test]
        async fn postgres() {
            assert_debug_snapshot!(engines::test::postgres(Search(SearchMode::FullText)).await);
        }
    }

    mod prefix {
        use atuin_client::settings::SearchMode;
        use insta::assert_debug_snapshot;

        use crate::command::client::search::engines::{self, db::Search};

        #[tokio::test]
        async fn docker_postgres() {
            assert_debug_snapshot!(engines::test::docker_postgres(Search(SearchMode::Prefix)).await);
        }

        #[tokio::test]
        async fn postgres() {
            assert_debug_snapshot!(engines::test::postgres(Search(SearchMode::Prefix)).await);
        }
    }
}
