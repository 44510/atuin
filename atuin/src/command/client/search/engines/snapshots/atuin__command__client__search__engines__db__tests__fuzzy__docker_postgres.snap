---
source: atuin/src/command/client/search/engines/db.rs
expression: "engines::test::docker_postgres(Search(SearchMode::Fuzzy)).await"
---
[
    History {
        id: "0",
        timestamp: 2023-06-16T11:14:16Z,
        duration: -1,
        exit: -1,
        command: "docker run -e POSTGRES_USER=atuin -e POSTGRES_PASSWORD=pass -e POSTGRES_DB=atuin -p 5432:5432 -d --rm postgres:14-alpine",
        cwd: "/Users/conrad/code/atuin",
        session: "1",
        hostname: "host1:conrad",
        deleted_at: None,
    },
]
